Fix inherents race condition

nazar-pc opened the pull request:
Fixes https://github.com/subspace/subspace/issues/871, see https://github.com/subspace/subspace/issues/871#issuecomment-1722745138 for the reason why this was happening.

Note that previously notification about importing block was technically happening AFTER block was imported (but before next block in import queue is imported, which wasn't the only way to get blocks imported), now it is happening BEFORE block is actually imported into Substrate.

So happy to finally understand what is going on there :relieved: 

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

nazar-pc committed:
Remove unnecessary field from `SubspaceBlockImport` because it is also present in `SubspaceLink`

nazar-pc committed:
Fix race conditions related to inherents and remove now unnecessary workaround from slot worker

nazar-pc committed:
Tiny cleanup

NingLin-P reviewed (commented):
Make sense overall! Plz also updates https://github.com/subspace/subspace/blob/1c9a7825267ecab281b97acbc9919f8b47d01645/test/subspace-test-service/src/lib.rs#L843 to reflect the change in domain tests.

nazar-pc committed:
Apply the same tweak to test service

rahulksnv commented on crates/sc-consensus-subspace/src/lib.rs:
```diff
@@ -1109,7 +1104,7 @@ where
         }
 
-        Ok(import_result)
+        self.inner.import_block(block).await
```
Is there some issue with sending the notification after this `self.inner.import_block(block).await` (except the time lag between import completion and sending the notification)? 

I am missing why it has to be sent before the import, which might actually fail afterwards (this being a behavior change if someone depends on the notification to indicate that import has finished successfully)

nazar-pc replied:
Did you read https://github.com/subspace/subspace/issues/871#issuecomment-1722745138?

The behavior chance is acceptable here, it was never really meant to be fired after block import or else notification would have been "imported", not "importing". And "imported" already exists, the only difference is it doesn't prevent block import from moving forward.

nazar-pc replied:
In fact in domains both "importing" and "imported" subscriptions are used together to achieve block import backpressure.

rahulksnv replied:
Thanks, now I understand the importing vs imported distinction. IIUC, the pre-processing that was supposed to happen before the import was happening out of order sometimes

nazar-pc committed:
Merge remote-tracking branch 'origin/main' into fix-inherents-race-condition

# Conflicts:
#	crates/sc-consensus-subspace/src/slot_worker.rs

nazar-pc committed:
Merge remote-tracking branch 'origin/main' into fix-inherents-race-condition

# Conflicts:
#	crates/sc-consensus-subspace/src/lib.rs

NingLin-P committed:
Ensure consensus block is fully processed by operator before next slot in test

This commit sending one more notification to the operator to ensure it have
received the previous consensus block import notification and ensure it have
finish processing it before return. This commit also clean up some workaround
introduced previously that no more necessary or is fixed in upstream.

Signed-off-by: linning <linningde25@gmail.com>

nazar-pc commented on test/subspace-test-service/src/lib.rs:
```diff
@@ -469,17 +469,31 @@ impl MockConsensusNode {
-                    self.acknowledgement_sender_subscribers.retain(|subscriber| !subscriber.is_closed());
-                }
-            }
+        while (acknowledgement_receiver.next().await).is_some() {
```
`()` here and in the other file are unnecessary

nazar-pc commented on test/subspace-test-service/src/lib.rs:
```diff
@@ -469,17 +469,31 @@ impl MockConsensusNode {
+        {
+            // Must drop `block_import_acknowledgement_sender` after the notification otherwise
+            // the receiver will block forever as there is still a sender not closed.
+            let value = (Default::default(), acknowledgement_sender);
```
This is the usage of `Default` that is hard to understand. When I'm reviewing this diff I don't see what that type is and why it is fine to send `Default::default()`. `T::default()` is much better in most cases.

NingLin-P replied:
It is updated to `T::default()`, some comments also been added to explain why it is okay to use the default value along with a TODO.

nazar-pc replied:
Nice, thank you!

nazar-pc reviewed (commented):
Some nits, let's see if CI passes

NingLin-P committed:
Apply review suggestions

Signed-off-by: linning <linningde25@gmail.com>
